Add Dinic's Maximum Flow Algorithm

BryanCruz opened the pull request:
# Add Dinic's Maximum Flow Algorithm

References:
* https://en.wikipedia.org/wiki/Dinic%27s_algorithm
* https://cp-algorithms.com/graph/dinic.html

RaoulLuque commented:
As mentioned in [PR 740, by @starovoid ]https://github.com/petgraph/petgraph/pull/740#issuecomment-2888556128), the algorithm does not seem to be improving upon the existing implementation when running the benchmarks. Indeed it seems to be performing worse.

This may be because the graphs from the benchmark have the same number of nodes as edges. However, in such a setting, Ford Fulkerson and Dinics should actually perform similarly. The results on my PC are however as follows:

```
dinics_bench            ... bench:      28,915.92 ns/iter (+/- 2,343.77)
ford_fulkerson_bench            ... bench:      17,511.19 ns/iter (+/- 1,822.12)
```

Which shows the decrease in performance. I also added a similar benchmark for both an increased number of edges, where the Dinic performs similarly poor. The bench setup is as follows:

```rust
#[bench]
fn dinics_bench_many_edges(bench: &mut Bencher) {
    static NODE_COUNT: usize = 1_001;
    let mut g: Graph<usize, usize> = Graph::new();
    let nodes: Vec<NodeIndex<_>> = (0..NODE_COUNT).map(|i| g.add_node(i)).collect();
    for j in [1, 2, 4, 5, 10, 20, 25, 50] {
        for i in 0..(NODE_COUNT - 1) / j  {
            g.add_edge(nodes[i], nodes[(i + 1) * j], 1);
        }
    }
    bench.iter(|| {
        let _flow = dinics(
            &g,
            NodeIndex::from(0),
            NodeIndex::from(g.node_count() as u32 - 1),
        );
    });
}
```

And the results are:
```
dinics_bench_many_edges ... bench:     355,747.67 ns/iter (+/- 14,674.08)
ford_fulkerson_bench_many_edges ... bench:     205,071.71 ns/iter (+/- 3,839.07)
```

BryanCruz commented:
Hi @RaoulLuque , thanks for pointing out the benchmark and for providing a larger bench test.

I reimplemented some parts of the algorithm and got it to work a lot better. 

The Dinics and Ford Fulkerson benches that were being compared contained graph instances that favored a Ford Fulkerson method, so I added some others that favors Dinics. I added the benchmark results in the PR description.  

I'll also check the Maximum Bipartite Matching PR soon.

RaoulLuque commented:
(one thing I can already say tho is that a docstring in the [algo docs format](https://github.com/petgraph/petgraph/blob/master/assets/guide/algodocs.md) should be added :) )

BryanCruz commented:
Made a fix in the algorithm execution, and added a Min S-T Cut implementation.

Will check the algo docs format.

BryanCruz commented:
Added docs in [algo docs format](https://github.com/petgraph/petgraph/blob/master/assets/guide/algodocs.md)

RaoulLuque commented on src/algo/maximum_flow/dinics.rs:
```diff
@@ -0,0 +1,335 @@
+    N::EdgeWeight: Sub<Output = N::EdgeWeight> + PositiveMeasure,
+{
+    let mut flow_increase = N::EdgeWeight::zero();
+    let mut edge_to = vec![None; network.node_count()];
```
```suggestion
    let mut edge_to = vec![None; network.node_bound()];
```
See previous comments

RaoulLuque commented on src/algo/maximum_flow/dinics.rs:
```diff
@@ -0,0 +1,335 @@
+
+/// Compute the maximum flow from `source` to `destination` in a directed graph.
+///
+/// Implements [Dinic's (or Dinitz's) algorithm][ff], which builds successive
+/// level graphs using breadth-first search and finds blocking flows within
+/// them through depth-first searches.
+///
+/// # Arguments
+/// * `network` — A directed graph with positive edge weights, namely "flow capacities".
+/// * `source` — The source node where flow originates.
+/// * `destination` — The destination node where flow terminates.
+///
+/// # Returns
+/// Returns a tuple of two values:
+/// * `N::EdgeWeight`: computed maximum flow;
+/// * `Vec<N::EdgeWeight>`: the flow of each edge. The vector is indexed by the graph's edge indices.
+///
+/// # Complexity
+/// * Time complexity:
+///   * In general: **O(|V|²|E|)**
+///   * In networks with only unit capacities: **O(min{|V|²ᐟ³, |E|¹ᐟ²} |E|)**
+/// * Auxiliary space: **O(|V| + |E|)**.
+///
+/// where **|V|** is the number of nodes and **|E|** is the number of edges.
+///
+/// [ff]: https://en.wikipedia.org/wiki/Dinic%27s_algorithm
```
```suggestion
/// Implements [Dinic's (or Dinitz's) algorithm][dinics], which builds successive
/// level graphs using breadth-first search and finds blocking flows within
/// them through depth-first searches.
///
/// # Arguments
/// * `network` — A directed graph with positive edge weights, namely "flow capacities".
/// * `source` — The source node where flow originates.
/// * `destination` — The destination node where flow terminates.
///
/// # Returns
/// Returns a tuple of two values:
/// * `N::EdgeWeight`: computed maximum flow;
/// * `Vec<N::EdgeWeight>`: the flow of each edge. The vector is indexed by the graph's edge indices.
///
/// # Complexity
/// * Time complexity:
///   * In general: **O(|V|²|E|)**
///   * In networks with only unit capacities: **O(min{|V|²ᐟ³, |E|¹ᐟ²} |E|)**
/// * Auxiliary space: **O(|V| + |E|)**.
///
/// where **|V|** is the number of nodes and **|E|** is the number of edges.
///
/// [dinics]: https://en.wikipedia.org/wiki/Dinic%27s_algorithm
```

RaoulLuque reviewed (commented):
Hey, thanks for reworking some things from the initial implementation and adapting the docs to match our algo docs :sauropod: 

I would suggest to move the minimum S-T Cut to another PR :+1: 

Other than that, I suggested a few changes, but in general, LGTM :)

I was wondering with the benchmarks, whether you looked at some of the graphs from the common.rs and took into consideration using one of those?

BryanCruz commented on src/algo/maximum_flow/dinics.rs:
```diff
@@ -0,0 +1,244 @@
+///
+/// [ff]: https://en.wikipedia.org/wiki/Dinic%27s_algorithm
+pub fn dinics<N>(
+    network: N,
```
> I was wondering with the benchmarks, whether you looked at some of the graphs from the common.rs and took into consideration using one of those?

I took a look at the `factories.rs`, but didn't found any weighted graph.

The tests for Maximum Flow should be a bit more specific than generic graphs because of the network nature the problem instance should have. Also, we need Source and Destination nodes

So its a lot easier to create specific instances for benchmarks than adapt some existing one 🤔 

> I am not sure about the choice of name of N as generic here, since N is already used as the name of the generic parameter for the node labels in the Graph struct. I imagine you've just copied the naming over from Ford Fulkerson, but I would prefer using G and am interested in what you think or whether I am missing something?

I don't have a strong preference here. I think `N` also makes sense because we're dealing with **N**etworks, but I can change to `G` in order to not conflict with node labels default generic name.

> If you agree, we could also change the generic parameter naming in Ford Fulkerson while we are at it (or in another small PR) :)

I can change here, no problem

BryanCruz replied:
> I would suggest to move the minimum S-T Cut to another PR 👍

#847

BryanCruz commented on src/algo/maximum_flow/dinics.rs:
```diff
@@ -0,0 +1,335 @@
+    let mut max_flow = N::EdgeWeight::zero();
+    let mut flows = vec![N::EdgeWeight::zero(); network.edge_count()];
+    let mut visited = network.visit_map();
+    let mut level_edges = vec![Default::default(); network.node_count()];
```
done

also added a test to check the correct behavior in stable graphs

RaoulLuque replied:
That's great, thanks :sun_with_face: 

This is just a general not-quite-satisfiedness of myself with the current testing, but probably in the future there should be some framework for us to test that algorithms that are supposed to work in different classes, actually do :) But that's just a side note and does not have to bother you ^^

BryanCruz commented on src/algo/maximum_flow/dinics.rs:
```diff
@@ -0,0 +1,335 @@
+    } else {
+        b
+    }
+}
```
instead of adding the `Ord` trait requirement, I added `panic` when neither `a < b` or `a >= b` are `true`

I think its worth to keep the `PartialOrd` to handle cases like floats as edge weights

RaoulLuque replied:
Ah, that is a good point. Allowing floats is definitely desirable. I am not sure though whether panicking is desirable.

Maybe someone else has a good idea for handling this, but maybe we should do if b < a { b } else { a } and then always call it with the existing value as a and the new value as b. This way, we won't update, if the values are not updatable?

BryanCruz replied:
I'm in favor of `panic!` because if the algorithm doesn't work for a certain graph instance, then the reason is very clear for the user.

Without raising any errors and failing silently, if values are not comparable for some reason, the algorithm could give an invalid answer to the user and they may not know why it's wrong or not even notice.

If someone realy needs the algorithm to work in graph instances with values that are not comparable, they can make a few tweaks and create another function like `dinics_unsafe`. Since I don't know if such case will ever exist, I wouldn't implement it now.

What do you think ?

RaoulLuque replied:
I see, yeah. You're arguments do make sense :+1: As I said I am torn, or rather fine with both solutions :) Just would need to document whichever of the two clearly in the docstring I :+1:

RaoulLuque commented:
Sry for the additional comment, but since you seem to be quite knowledgable when it comes to the different use-cases of Dinics/Ford Fulkerson (Edmonds Karp), I think it would be a good idea to document when to use which function. We could use the `maximum_flow/mod.rs` for this and add a comment there. In the `ford_fulkerson` and `dinics` docstrings we could then refer to the `mod.rs` docs for when to use this or other maximum flow functions :)

BryanCruz commented:
> I think it would be a good idea to document when to use which function. We could use the `maximum_flow/mod.rs` for this and add a comment there. In the `ford_fulkerson` and `dinics` docstrings we could then refer to the `mod.rs` docs for when to use this or other maximum flow functions :)

good idea, added in latest commit

RaoulLuque commented:
Okay, so if you adjust the docstring of dinics to document the possibility to panic as stated in https://github.com/petgraph/petgraph/pull/739#discussion_r2217908348, I think this Pr would be ready to merge, thanks a lot again :star2: 

Just one more comment I have, just below. Note that you can also accept suggestions right here on GitHub, this way the reviewers will be credited in Git for their suggestions :)

RaoulLuque commented on src/algo/maximum_flow/ford_fulkerson.rs:
```diff
@@ -0,0 +1,215 @@
+    }
+}
+
+/// [Ford-Fulkerson][ff] algorithm in the [Edmonds-Karp][ek] variation.
+///
+/// Computes the [maximum flow] from `source` to `destination` in a weighted directed graph.
```
```suggestion
/// [Ford-Fulkerson][ff] algorithm in the [Edmonds-Karp][ek] variation.
/// Computes the [maximum flow] from `source` to `destination` in a weighted directed graph.
```
This makes it such that on the module docs for `maximum_flow`, the description of Ford Fulkerson also shows the text about computing a max flow :)

RaoulLuque commented on src/algo/maximum_flow/dinics.rs:
```diff
@@ -310,6 +314,8 @@ where
 
 /// Returns the minimum value between given `a` and `b`.
+/// Will panic if it tries to compare two elements that aren't comparable
+/// (i.e., given two elements `a` and `b`, neither `a >= b` or `a < b`).
```
```suggestion
/// (i.e., given two elements `a` and `b`, neither `a >= b` nor `a < b`).
```

RaoulLuque commented:
Looks great, thanks for all your patience with my review :relaxed:  Just these two small language nitpicks, after which I will merge the PR :+1:

BryanCruz commented:
> Looks great, thanks for all your patience with my review ☺️ Just these two small language nitpicks, after which I will merge the PR 👍

thanks ☺️ , commited the latest suggestions
